### PR TITLE
raftstore-v2: publish tablet in raftstore thread only

### PR DESCRIPTION
**Pull request by BusyJay**
### What is changed and how it works?
Issue Number: Ref #12842 

What's Changed:

```commit-message
Publish tablet in apply thread is unsafe. This PR moves the operation to
raftstore. It also fixes the issues that applying two splits at a time can
cause panic. It also makes sure cache will be cleared after tablet is published.
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore-v2/src/worker/tablet_gc.rs [hunk 1]
```diff
@@ -156,10 +156,15 @@ impl<EK: KvEngine> Runner<EK> {
                 "path" => path.display(),
             ),
             Ok(false) => {
+                let (_, region_id, tablet_index) =
+                    registry.parse_tablet_name(path).unwrap_or(("", 0, 0));
                 // TODO: use a meaningful table context.
```
Comment (tabokie) [inline components/raftstore-v2/src/worker/tablet_gc.rs:161]: remove?

File: components/raftstore-v2/src/raft/peer.rs [hunk 2]
```diff
@@ -346,41 +336,18 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
-                .schedule(tablet_gc::Task::prepare_destroy(
-                    old_tablet.clone(),
-                    self.region_id(),
-                    new_tablet_index,
-                ));
-        }
-        // TODO: Handle race between split and snapshot. So that we can assert
-        // `self.tablet.refresh() == 1`
-        assert!(self.tablet.refresh() > 0);
+    pub fn reset_tablet(&mut self, tablet: EK) -> Option<EK> {
```
Comment (tabokie) [inline components/raftstore-v2/src/raft/peer.rs:339]: set_tablet seems more consistent.

File: components/raftstore-v2/src/operation/command/admin/compact_log.rs [hunk 3]
```diff
@@ -260,18 +346,26 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
+        context.approximate_raft_log_size = (context.approximate_raft_log_size as f64
+            * (remain_cnt as f64 / total_cnt as f64))
+            as u64;
     }
 
+    /// Called when apply index is persisted. There are two different situation:
+    ///
+    /// Generally, additional writes are triggered to persist apply index. In
+    /// this case task is `Some`. But after applying snapshot, the apply
+    /// index is persisted a head of time. In this case task is `None`.
```
Comment (tabokie) [inline components/raftstore-v2/src/operation/command/admin/compact_log.rs:362]:
```suggestion
    /// index is persisted ahead of time. In this case task is `None`.
```

File: components/raftstore-v2/src/operation/command/admin/compact_log.rs [hunk 4]
```diff
@@ -302,6 +402,10 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
         let truncated = self.entry_storage().truncated_index() + 1;
         let persisted_applied = self.storage().apply_trace().persisted_apply_index();
         let compact_index = std::cmp::min(truncated, persisted_applied);
+        if compact_index == RAFT_INIT_LOG_INDEX + 1 {
+            // There is no logs at RAFT_INIT_LOG_INDE, nothing to delete.
```
Comment (tabokie) [inline components/raftstore-v2/src/operation/command/admin/compact_log.rs:406]:
```suggestion
            // There is no logs at RAFT_INIT_LOG_INDEX, nothing to delete.
```

File: components/raftstore-v2/src/operation/query/local.rs [hunk 5]
```diff
@@ -50,6 +49,78 @@ where
     }
 }
 
+pub type ReadDelegatePair<EK> = (ReadDelegate, SharedReadTablet<EK>);
+
+/// A share struct for local reader.
+///
+/// Though it looks like `CachedTablet`, but it only stores one version of
```
Comment (tabokie) [inline components/raftstore-v2/src/operation/query/local.rs:56]: Not easy to understand. Maybe something like "Allow weak referencing a tablet. The initial instance of `SharedReadTablet` controls the lifetime of tablet. Subsequent clones can only temporarily borrow it if the initial instance is still alive".

**Review by tabokie — APPROVED** [review]

File: components/raftstore-v2/src/worker/tablet_gc.rs (hunk 1 above)
Comment (BusyJay) [reply components/raftstore-v2/src/worker/tablet_gc.rs:161]: It's not meaningful yet. For example, key range is missing.

**Comment by BusyJay** [comment]
PTAL

**Review by tonyxuqqi — APPROVED** [review]

**Comment by BusyJay** [comment]
//merge

**Pull request merged**